[WIP] randint function

Naman-ntc opened the pull request:
Added randint functionality as requested in [[#5874]](https://github.com/pytorch/pytorch/issues/5874).

Naman-ntc committed:
added randint function in ATEN yaml as well as Tensorfactories.cpp

SsnL commented:
This needs documentation in `_torch_docs.py` and tests in `test_torch.py`.

SsnL commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -249,6 +249,11 @@ Tensor& randperm_out(Tensor& result, int64_t n, Generator* generator) {
 
+Tensor randint(const Type& dtype, IntList size, Generator* generator) {
+  Tensor result = dtype.tensor(size);
+  return result.random_(0, 1, generator);
```
`randint` should take a range....

Naman-ntc replied:
Oops! commited wrong file, so sorry, my bad!! Will update and commit.

Naman-ntc committed:
corrected randint

Naman-ntc commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -249,6 +249,11 @@ Tensor& randperm_out(Tensor& result, int64_t n, Generator* generator) {
   return result;
 }
 
+Tensor randint(const Type& dtype, IntList size, int64_t low, int64_t high, Generator* generator) {
```
Yeah sure! Though I didnt understand their exact use but as for implementation it's trivial!

Naman-ntc replied:
Can you point me to their use just for completeness? I didnt ever saw use of randn_like or randn_out as well.

SsnL replied:
>>>  I didnt ever saw use of `randn_like`

It's not in a release yet, but it is available on master. So is other methods like `empty_like` :)

>>> 'Oh is the out parameter handled via rand_out?'

Yes

colesbury commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -501,6 +501,9 @@
 - func: randperm_out(Tensor result, int64_t n, *, Generator* generator=nullptr) -> Tensor
   variants: function
 
+- func: randint(Type dtype, IntList size, int64_t low, int64_t high, *, Generator* generator=nullptr) -> Tensor
```
This should follow the same argument ordering and conventions as numpy randint:

https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html

Naman-ntc replied:
Oh okay!

Naman-ntc committed:
randint with overloading complete,getting tuple of ints behaviour though

Naman-ntc committed:
done randintlike and randint_out

Left : adding docs and test, and remove the bug on size = (5)

Naman-ntc commented:
These are the behaviours on corner cases otherwise working fine :  

> torch.randint(5,(3))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: randint(): argument 'size' (position 2) must be tuple of ints, not int

> torch.randint(5,3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: randint(): argument 'size' (position 2) must be tuple of ints, not int

> torch.randint(5,10,3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: randint() received an invalid combination of arguments - got (int, int, int), but expected one of:
(int high, tuple of ints size, torch.Generator generator, torch.dtype dtype, int device, bool requires_grad)
(int low, int high, tuple of ints size, torch.Generator generator, torch.dtype dtype, int 
device, bool requires_grad)


The first outcome is bothering me. Any fix? (Others are fine in my opinion)

ezyang commented:
@pytorchbot test this please

The behavior in the first case is just a Python-ism. `(3)` is the same as `3`; the one element tuple is `(3, )`.

Naman-ntc commented:
@ezyang yeah that case is working (I should have included it in my examples), but torch.randn((3)) works right. So therefore torch.randint((3,5,(4)) must also work for sake of continuity right?

Naman-ntc committed:
Removed my error messages, ThRandomTensor will handle all exceptions

ezyang commented:
@pytorchbot test this please

ezyang commented:
So, perhaps what you are suggesting is that randint should have an option `randint(low, high, *args)`, where args are sizes?

Naman-ntc commented:
@ezyang No we need to pass the size as a tuple only as we provide
randint(5,(2,2)) tu give random integers in range 0-5
If we dont enclose size in tuple then we cant differentiate between randint(5,2,(2)) and randint(5,(2,2))

ezyang commented:
No, what I am saying is that `(3)` is not a tuple of size one.

```
>>> (3)
3
>>> (3,)
(3,)
```

It accidentally works in `torch.randn((3)) ` case because randn has an overload `torch.randn(*size_args)`.

Best not to get too crazy with the overloads here, IMO.

Naman-ntc commented:
Ohh i see!! So i believe currently it's working fine, as we cant provide size_args overloading coz it'll complicate things. Will wrote the documentation and tests then. 
Will mention that for tensor of size 3 use sizes as (3,)

By the way if I overload it as torch.randint(int,int,int) then torch.randint(1,5,(3)) would work right. I can do that if you want.

ezyang commented:
I hate it! Leave it the way it is ;)

SsnL commented:
Looks reasonable, but this still needs doc and tests.

Naman-ntc commented:
Will commit by tommorow @SsnL  :)

Naman-ntc commented:
I am having doubt how should I document the function.
As I have used function overloading I can create a function signature like this : 
 
`torch.randint(low=0,high,sizes)` (low can be optional)
Now I think it might confuse some people as it's not possible for first argument to be optional and others not, do you think it's alright to document it like this??

SsnL commented:
We already have docs where first arg is written as optional, e.g. http://pytorch.org/docs/master/torch.html#torch.arange. So it should be fine.

Naman-ntc committed:
added docs and tests, corrected a mistake

Tested with manual seeds in some test cases as well. Seems fine to me (check documentation though)

Naman-ntc committed:
corrected indentation to spaces, and improved sizes argument description

Naman-ntc committed:
made documentation argument description shorter

Naman-ntc committed:
added whitespace after ',' in torch docs

Naman-ntc commented:
@SsnL I guess it's ready to merge now?

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
+
+Example::
+
+    >>> torch.randint(3,5,(3,))
```
nit: spaces after `,`

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
+     3
+    [torch.FloatTensor of size (3,)]
+
+    >>> torch.randint(5,(2,2))
```
spaces here as well

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
+     0  2
+    [torch.FloatTensor of size (2,2)]
+
+    >>> torch.randint(3,10,(2,2))
```
spaces here as well

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
+The shape of the tensor is defined by the variable argument :attr:`sizes`.
+
+Args:
+    low (int): Lowest (positive) integer to be drawn from the distribution.
```
nit: `int, optional`. Should also have ``Default: `0` `` at the end.

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
 
+add_docstr(torch.randint,
+           r"""
+randn(low=0,high,sizes, out=None) -> Tensor
```
nit spaces

SsnL commented on test/test_torch.py:
```diff
@@ -2549,6 +2549,14 @@ def test_randn(self):
         self.assertEqual(res1, res2)
 
+    def test_randint(self):
+        torch.manual_seed(123456)
```
Can we test all possible combinations, i.e., `torch.randint(high, size)`, `torch.randint(low, high, size)` and the out variants? The above `test_rand` and `test_randn` are not the best examples... Sorry about those. 

Additionally, can you also check that the sampled values are within bounds?

Naman-ntc replied:
Any particular function I should use for checking bounds or simple combination of < > operators and summing up tensor?

SsnL replied:
Anything that is reasonable to you is good :)

Naman-ntc committed:
addes spaces in documentation

Naman-ntc committed:
added more tests (including bounds and overloading features)

Naman-ntc committed:
added whitespaces in test_torch

Naman-ntc committed:
removed trailing whitespaces

Naman-ntc committed:
removed whitespace from a blank line

Naman-ntc commented:
@SsnL is there anything more to add or correct now?

SsnL commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
+The shape of the tensor is defined by the variable argument :attr:`sizes`.
+
+Args:
+    low (int, optional): Lowest (positive) integer to be drawn from the distribution. Default: 0.
```
Any reason that `low` and `high` need to be positive? `0` is not positive as well.

Naman-ntc replied:
Yeah so in the develop mode 
> >>> a = torch.randn(5,6)
>>> a.random_(-10,-1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/naman/PythonPackages/Source-Pytorch/lib/python3.6/site-packages/torch/autograd/variable.py", line 84, in __repr__
    return torch._tensor_str._str(self)
  File "/home/naman/PythonPackages/Source-Pytorch/lib/python3.6/site-packages/torch/_tensor_str.py", line 306, in _str
    strt = _matrix_str(self)
  File "/home/naman/PythonPackages/Source-Pytorch/lib/python3.6/site-packages/torch/_tensor_str.py", line 215, in _matrix_str
    min_sz=5 if not print_full_mat else 0)
  File "/home/naman/PythonPackages/Source-Pytorch/lib/python3.6/site-packages/torch/_tensor_str.py", line 95, in _number_format
    if value != math.ceil(value.item()):
RuntimeError: Overflow when unpacking long

So I thought random_ works only for positive numbers. But  just now checked it's overflow problem and works for torch.LongTensor. 
By the way isn't it random_ implementation issue. In default settings for FloatTensors or even DoubleTensors it would generate this message? 
I'll remove positive but I should add warning of overflow then maybe?

SsnL replied:
@li-roy See print error above.

SsnL replied:
@Naman-ntc Let's remove "positive". The error you saw was purely print format error. The actual method works on nonpositive values.

Naman-ntc replied:
@SsnL removed in next commit, but still randint wont work in FloatTensor or DoubleTensor setting right (for negatives)! Is it alright, because if he needs DoubleTensor randint user'll have to typecase it to LongTensor and then to DoubleTensor which is very 'smooth' imo?

SsnL replied:
That is an related bug and shouldn't affect your PR. I opened an issue at #6338

Naman-ntc replied:
Great Thanks a lot :+1:

Naman-ntc replied:
@SsnL  how big would be a patch for #6338 ? If it's not too big I would like try it over a few days.

SsnL commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -495,6 +495,30 @@
 - func: randn_like(Tensor self, *, Type dtype) -> Tensor
   variants: function
 
+- func: randint(Type dtype, int64_t high, IntList size, *, Generator* generator=nullptr) -> Tensor
```
nit: can you move this above `randn*` so that this file is alphabetically ordered.

ssnl reviewed (commented):
Generally looks great except for two minor things.

fmassa commented on torch/_torch_docs.py:
```diff
@@ -4108,6 +4108,44 @@
 
+add_docstr(torch.randint,
+           r"""
+randint(low=0, high, sizes, out=None) -> Tensor
```
Can't we pass a dtype object to randint?

Naman-ntc replied:
Yeah, I just tried to mimic randn and rand in documentation which also expect type argument (in ATEN) but don't mention it in documentation. (Probably take default tensor set in environment)
Should I add it in doc?

SsnL replied:
Add it please. The doc for `randn` and `rand` needs to be and will be updated.

Naman-ntc replied:
Alright, Default is torch.float32 right?

Naman-ntc committed:
removed positive requirement from docs. Added dtype argument and gave eg

Naman-ntc committed:
made randint over randn in all files

ssnl reviewed (approved):
LGTM, but need some owner to look at this.

fmassa commented on torch/_torch_docs.py:
```diff
@@ -4073,6 +4073,45 @@
+    high (int): One above the highest integer to be drawn from the distribution.
+    sizes (tuple): a tuple defining the shape of the output tensor.
+    out (Tensor, optional): the output tensor
+    dtype (torch.dpython:type, optional) – the desired type of returned Tensor. Default: torch.float32
```
This looks like a typo?

Naman-ntc replied:
Didnt get you? Are you talkin about torch.dpython:type ??

fmassa replied:
yes, is that expected?

Naman-ntc replied:
In ATEN :
`Tensor randint(const Type& dtype, int64_t low, int64_t high, IntList size, Generator* generator)`
Found similar parameter in Reference : [torch.bartlett_window](http://pytorch.org/docs/master/torch.html#torch.bartlett_window) so copied it! 
Not Sure though!

fmassa replied:
the link you posted doesn't render great, so I think it's not right. @SsnL do you know what's the right way of doing it?

SsnL replied:
hmm `bartlett_window` uses `torch.dtype`(https://github.com/pytorch/pytorch/blob/master/torch/functional.py#L281). But that get's incorrectly rendered to this somehow. I'll try to fix it.

Btw, the ` - ` should be a colon.

Naman-ntc committed:
changed to data type for dtype in docs for randint

Naman-ntc commented:
Can we merge this now?

colesbury reviewed (approved):
Thanks!

vadimkantorov commented:
`randint` doesn't show up in site docs: http://pytorch.org/docs/master/search.html?q=randint&check_keywords=yes&area=default

Naman-ntc commented:
It'll probably come in next release!

vadimkantorov commented:
I think some `autofunction` entry is needed, so that Sphinx would pull up the docs from the doc string

fmassa commented:
Looks like @vadimkantorov is right.
@Naman-ntc can you add an entry for `randint` in https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst ? Probably something like
```
.. autofunction:: randint
```
in the random functions section, look [here](https://raw.githubusercontent.com/pytorch/pytorch/master/docs/source/torch.rst)

Naman-ntc commented:
Ohh I see!
Makes sense, will make a new PR for it.
